app/src/main/java/org/mozilla/fenix/components/metrics/ActivationPing.kt

```
@@ -0,0 +1,187 @@
+            // within a single product, but we don't want hashes to be the same across different
+            // products (e.g. Fennec vs Fenix).
+            val salt = getHashingSalt()
+
+            // Apply hashing.
+            try {
+                // Note that we intentionally want to use slow hashing functions here in order
+                // to increase the cost of potentially repeatedly guess the original unhashed
+                // identifier.
+                val keySpec = PBEKeySpec(
+                    unhashedID.toCharArray(),
+                    salt.toByteArray(),
+                    PBKDF2_ITERATIONS,
+                    PBKDF2_KEY_LEN_BITS)
+
+                val keyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1")
+                val hashedBytes = keyFactory.generateSecret(keySpec).encoded
+                Base64.encodeToString(hashedBytes, Base64.NO_WRAP)
+            } catch (e: java.lang.NullPointerException) {
+                Logger.error("ActivationPing - missing or wrong salt parameter")
+                null
+            } catch (e: IllegalArgumentException) {
+                Logger.error("ActivationPing - wrong parameter", e)
+                null
+            } catch (e: NoSuchAlgorithmException) {
+                Logger.error("ActivationPing - algorithm not available")
+                null
+            } catch (e: InvalidKeySpecException) {
+                Logger.error("ActivationPing - invalid key spec")
+                null
+            }
+        }
+    }
+
+    /**
+     * Fills the metrics and triggers the 'activation' ping.
+     * This is a separate function to simplify unit-testing.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun triggerPing() {
+        // Generate the activation_id.
+        Activation.activationId.generateAndSet()
+
+        CoroutineScope(Dispatchers.Default).launch {
```

Note that calling launch without joining on its job will fire-and-forget. Exceptions will be swallowed and there's no guarantee it ever completes. This might be your intent, but you should be aware.

---

Thanks, yes, this is intended behaviour: we should be fine to swallow exceptions here and re-attempt, if something went wrong, next time we start.